pkg/segment/reader/metrics/tagstree/tagstreereader.go

```diff
@@ -180,6 +181,20 @@ func (attr *AllTagTreeReaders) FindTSIDS(mQuery *structs.MetricsQuery) (*tsidtra
 			rawTagValueToTSIDs := make(map[string]map[uint64]struct{})
 			for {
 				_, tagRawValue, tsids, tagRawValueType, more := itr.Next()
+
+				// if operator is regex check for match and skip on no match
+				if tf.TagOperator == segutils.Regex && len(tagRawValue) > 0 {
+					full_anchor_match := fmt.Sprintf("^(%v)$", tagVal)
```

Comment:
Also do this anchoring of Tag Value at the parser side. So that you would not do it for every segement.

And check whether the given regex is a valid regex or not using `regexp.Compile(pattern)`. So that if there is an error with the regex pattern that the user has given, we would reject it at the start itself.

Reply:
For an incorrect regex it seems that we directly return from parsePromQLQuery because ParseExpr (promql parser) reports an error and then the flow exits and ApplyMetricsQuery is not called.

Reply:
Okay got it.